core/src/main/java/org/fao/geonet/kernel/security/openidconnect/SimpleOidcUser.java

```diff
@@ -53,28 +54,67 @@ class SimpleOidcUser {
     private Address address;
 
 
-    SimpleOidcUser(OIDCConfiguration oidcConfiguration, OIDCRoleProcessor oidcRoleProcessor, OidcIdToken idToken) {
+    /**
+     *
+     * @param oidcConfiguration OIDC Configuration (mostly controlled by environment vars)
+     * @param oidcRoleProcessor Processes roles from the ID Token
+     * @param idToken  The User's ID token
+     * @param attributes All the user's claims (ID Token claims + USERINFO claims)
+     */
+    SimpleOidcUser(OIDCConfiguration oidcConfiguration, OIDCRoleProcessor oidcRoleProcessor, OidcIdToken idToken, Map attributes) {
+        if (attributes == null) {
+            attributes = new HashMap(); // easier if we remove all the null checks
+        }
+
+        // -- get user name.  Should be in ID Token, but could be in the USERINFO
         username = idToken.getPreferredUsername();
         if (username == null) {
             username = idToken.getFullName();
         }
+        if ( (username == null) && attributes.containsKey(StandardClaimNames.PREFERRED_USERNAME) ) {
```

Comment:
One issue that could exists in this block is that once the username is identified base on one of these fields then any changes could end up creating a new user and the more fields we base the username on, the more it amplifies this issue.

i.e.  I have seen an issues in the past where an organization had mapped AD to AAD (with case sensitivity) so all emails/preferred emails contained both cases in the email address.  Then they decided to change it so it was lowercase and this caused all existing users to lose their information as they were now treated as a new user. i.e. old user Test.User@example.com  and new user test.user@example.com.   

The point I'm trying to make is not so much about the issue that I witnessed.   But more regarding the issue that the more fields that we use to derive the username from - the more chances can arise when changes are made.
i.e. Assuming that someone configures the system and there is no email/preferred_username so the NAME is used.  Then they realize later they should have set the email address (Opps all users loose access to their data) - and then later they noticed that EMAIL contain case and they would prefer to use lowercase the preferred_username (Opps all users loose access to their data)

It would be better if we could configure the system to only use one location. 

I will not reject the PR based on this issue but it is something to consider.

Reply:
I modelled this code base on this:
 https://github.com/geonetwork/core-geonetwork/blob/main/core/src/main/java/org/fao/geonet/kernel/security/keycloak/KeycloakUserUtils.java#L340-L343
(i.e. how the keycloak plugin did this)

I am fine throwing an exception if `email/preferred_username` isn't present.

Reply:
I added a commit for this - it goes the configured claim, preferred_username, then email.  Otherwise it throws an exception.
It should really throw an exception if the configured claim doesn't give an acceptable value.